src/student_success_tool/modeling/bias_detection.py

```diff
@@ -0,0 +1,213 @@
+import pandas as pd
+import numpy as np
+import scipy.stats as st
+from sklearn.metrics import confusion_matrix
+
+# FNPR sample threshold
+MIN_FNPR_SAMPLES = 50
+Z = st.norm.ppf(1 - (1 - 0.95) / 2)
+
+# Flag FNPR difference thresholds
+HIGH_FLAG_THRESHOLD = 0.15
+MODERATE_FLAG_THRESHOLD = 0.1
+LOW_FLAG_THRESHOLD = 0.05
+
+# Define flag types
+FLAG_NAMES = {
+    "🟢 NO BIAS": "no_bias",
+    "🟡 LOW BIAS": "low_bias",
+    "🟠 MODERATE BIAS": "moderate_bias",
+    "🔴 HIGH BIAS": "high_bias",
+}
+
+
+def calculate_fnpr_and_ci(
+    targets: pd.Series,
+    preds: pd.Series,
+    min_fnpr_samples: int = MIN_FNPR_SAMPLES,
+) -> tuple[float, float, float]:
+    """
+    Calculates the False Negative Prediction Rate (FNPR) and its confidence interval.
+
+    Args:
+        targets: Labels from model output
+        preds: Predictions from model output
+        min_fnpr_samples: Minimum number of true positives or false negatives for FNPR calculation. When TP or FN are low, FNPR can be very unstable and flag subgroups based on small differences in TP or FN.
+    Returns:
+        Tuple of FNPR, lower CI bound, and upper CI bound.
+    """
+    cm = confusion_matrix(targets, preds, labels=[False, True])
+    tn, fp, fn, tp = (
+        cm.ravel()
+        if cm.shape == (2, 2)
+        else np.bincount(np.array(targets) * 2 + np.array(preds), minlength=4)
+    )
+
+    denominator = fn + tp
+
+    if (tp < min_fnpr_samples) or (fn < min_fnpr_samples):
+        return np.nan, np.nan, np.nan
+
+    fnpr = fn / denominator
+    margin = Z * np.sqrt((fnpr * (1 - fnpr)) / denominator)
+    return fnpr, max(0, fnpr - margin), min(1, fnpr + margin)
+
+
+def check_ci_overlap(
+    ci1: tuple[float, float],
+    ci2: tuple[float, float],
+) -> bool:
+    """
+    Checks whether confidence intervals (CIs) overlap. If they do, the FNPR differences
+    are within the margin of error at the 95% confidence level. If the CIs do not
+    overlap, this suggests strong statistical evidence that the FNPRs are different.
+
+    Args:
+        ci1: Confidence interval (min, max) for subgroup 1
+        ci2: Confidence interval (min, max) for subgroup 2
+    Returns:
+        Boolean indicating whether the CIs overlap.
+    """
+    return not (ci1[1] < ci2[0] or ci2[1] < ci1[0])
+
+
+def z_test_fnpr_difference(
+    fnpr1: float,
+    fnpr2: float,
+    denominator1: int,
+    denominator2: int,
```

Comment:
"denominator" is vague, I think this would be clearer with variables named like `num_negatives` or something